api/src/npm/specifiers.rs

```diff
@@ -1,12 +1,105 @@
 // Copyright 2024 the JSR authors. All rights reserved. MIT license.
+
+use std::collections::HashMap;
+
+use deno_ast::ModuleSpecifier;
+use deno_graph::Dependency;
 use deno_semver::jsr::JsrPackageReqReference;
 use deno_semver::npm::NpmPackageReqReference;
+use indexmap::IndexMap;
 
 use crate::ids::ScopedPackageName;
 
 use super::NpmMappedJsrPackageName;
 
-pub fn rewrite_specifier(specifier: &str) -> Option<String> {
+#[derive(Clone, Copy)]
+pub enum RewriteKind {
+  Source,
+  Declaration,
+}
+
+#[derive(Clone, Copy)]
+pub struct SpecifierRewriter<'a> {
+  pub base_specifier: &'a ModuleSpecifier,
+  pub source_rewrites: &'a HashMap<&'a ModuleSpecifier, ModuleSpecifier>,
+  pub declaration_rewrites: &'a HashMap<&'a ModuleSpecifier, ModuleSpecifier>,
+  pub dependencies: &'a IndexMap<String, Dependency>,
+}
+
+impl<'a> SpecifierRewriter<'a> {
+  pub fn rewrite(&self, specifier: &str, kind: RewriteKind) -> Option<String> {
```

Comment:
would be nice to have a unit test

Reply:
This is very difficult to unit test at a level lesser than fully integrated into the publishing flow where we have a proper `deno_graph::ModuleGraph`. This is because the rewriter needs to have a valid list of resolved dependencies for the given module in the form of `Dependency` structs. These are very complicated to manually write, as they have source ranges attached.

The spec test with the .txt files is already a unit test that exercises this code, so I don't want to spend significant additional work to test this function individually. I don't think it adds much.